### PR TITLE
Properly handle parent modules w/ parameters in `BaseFinetuning` callback

### LINKED ISSUE #7930: Parent (i.e. non-leaf) Modules w/ Parameters Not Properly Handled By Finetuning Callback
## 🐛 Bug

Parent (i.e. non-leaf) modules that have parameters themselves (rather than just their nested submodules having parameters) are not properly handled by several finetuning callback methods. The parent module parameters are ignored by all BaseFineTuning methods that depend upon BaseFineTuning.flatten_modules. I'll add a PR to address the issue shortly but am including a test to replicate the issue below. As a practical example of the issue, I initially encountered this bug when using BaseFinetuning w/ deberta, specifically, the [DisentangledSelfAttention](https://github.com/huggingface/transformers/blob/f8bd8c6c7eef0a7280cd58f89016ede5b77f142f/src/transformers/models/deberta/modeling_deberta.py#L502) parent module

#### New test to reproduce (tests/callbacks/test_finetuning_callback.py):
```python
def test_parent_module_w_param_model():
    """Test flattening, freezing, and thawing of models which contain parent (non-leaf) modules with parameters
    directly themselves rather than exclusively their submodules containing parameters.
    """
    class ConvBlock(nn.Module):

        def __init__(self, in_channels, out_channels):
            super().__init__()
            self.conv = nn.Conv2d(in_channels, out_channels, 3)
            self.act = nn.ReLU()
            self.bn = nn.BatchNorm2d(out_channels)

        def forward(self, x):
            x = self.conv(x)
            x = self.act(x)
            return self.bn(x)

    class ConvBlockParam(nn.Module):

        def __init__(self, in_channels, out_channels):
            super().__init__()
            self.conv = nn.Conv2d(in_channels, out_channels, 3)
            self.act = nn.ReLU()
            # add trivial test parameter to conv block to validate parent (non-leaf) module parameter handling
            self.parent_param = nn.Parameter(torch.zeros((1), dtype=torch.float))
            self.bn = nn.BatchNorm2d(out_channels)

        def forward(self, x):
            x = self.conv(x)
            x = self.act(x)
            return self.bn(x)

    model = nn.Sequential(
        OrderedDict([
            ("encoder", nn.Sequential(ConvBlockParam(3, 64), ConvBlock(64, 128))),
            ("decoder", ConvBlock(128, 10)),
        ])
    )

    # There are 10 leaf modules or parent modules w/ parameters in the test model
    assert len(BaseFinetuning.flatten_modules(model)) == 10

    BaseFinetuning.freeze(model.encoder, train_bn=True)
    assert not model.encoder[0].conv.weight.requires_grad  # Validate a leaf module parameter is frozen
    assert not model.encoder[0].parent_param.requires_grad  # Validate the parent module parameter is frozen
    assert model.encoder[0].bn.weight.requires_grad

    BaseFinetuning.make_trainable(model)
    encoder_params = list(BaseFinetuning.filter_params(model.encoder, train_bn=True))
    # The 9 parameters of the encoder are:
    # conv0.weight, conv0.bias, bn0.weight, bn0.bias, parent_param
    # conv1.weight, conv1.bias, bn1.weight, bn1.bias
    assert len(encoder_params) == 9
```
### Expected behavior

parent_param in the above example model with nested modules should be appropriately handled by BaseFinetuning.[flatten_modules, freeze, make_trainable, filter_params] instead of omitted.

### Environment
* CUDA:
        - GPU:
                - GeForce RTX 2070 SUPER
                - GeForce RTX 2070
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.20.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.1
        - pytorch-lightning: 1.4.0dev
        - tqdm:              4.61.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #80-Ubuntu SMP Mon Apr 12 17:35:00 UTC 2021

### Additional context

I've got a fix ready and will be submitting a PR shortly. Thanks for all the great work on this awesome framework!

### PR DESCRIPTION
**Pull request by speediedan**
## What does this PR do?
Fixes #7930

This PR fixes issue #7930 and provides relevant test coverage: test_finetuning_callback.py::test_parent_module_w_param_model

Models w/ parent modules that directly contained parameters themselves (as opposed to parameters exclusively in their constituent submodules) had their parameters ignored by the BaseFinetuning callback. The primary issue was addressed with an additional filtering condition in BaseFinetuning.flatten_modules(). To avoid passing duplicate parameters in methods that depended upon BaseFinetuning.flatten_modules, I also added recurse=False to the parameter loops in BaseFinetuning.[freeze, make_trainable, filter_params]. The new test I've added could potentially be consolidated with /tests/callbacks/test_finetuning_callback.py::test_deep_nested_model but I at least initially have broken the test out separately for clarity. As a practical example of the issue, I initially encountered this bug when using BaseFinetuning w/ deberta, specifically, the [DisentangledSelfAttention](https://github.com/huggingface/transformers/blob/f8bd8c6c7eef0a7280cd58f89016ede5b77f142f/src/transformers/models/deberta/modeling_deberta.py#L502) parent module

I've updated the docstring of flatten_modules() to reflect the changes. All test_finetuning_callback.py tests are passing on my end.

- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃
YES! This is my first contribution to PL but I'm hoping to be more involved in the future. Thanks for the awesome product/work!

### CODE REVIEW

File: tests/callbacks/test_finetuning_callback.py [hunk 1]
```diff
@@ -342,3 +342,59 @@ def forward(self, x):
     # conv0.weight, conv0.bias, bn0.weight, bn0.bias
     # conv1.weight, conv1.bias, bn1.weight, bn1.bias
     assert len(encoder_params) == 8
+
+
+def test_parent_module_w_param_model():
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:347]: >  but I at least initially have broken the test out separately for clarity. 

It is fine to replace the one above as you are just adding checks! So please do

**Review by carmocca — COMMENTED** [review]
Nice catch!

Remember to update the CHANGELOG.md file

File: tests/callbacks/test_finetuning_callback.py (hunk 1 above)
Comment (speediedan) [reply tests/callbacks/test_finetuning_callback.py]: Great idea, thanks. I've added another commit that replaces test_deep_nested_model with the new test and changes the test name to be more general as I suspect there may be other minor checks we will want to add for more complex models that may not warrant separate tests. Also updated changelog. Thanks for the great framework! Looking forward to contributing in any meager way I can.

File: tests/callbacks/test_finetuning_callback.py [hunk 2]
```diff
@@ -307,7 +307,10 @@ def configure_optimizers(self):
     trainer.fit(model)
 
 
-def test_deep_nested_model():
+def test_complex_nested_model():
+    """Test flattening, freezing, and thawing of models which contain parent (non-leaf) modules with parameters
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:311]:
```suggestion
    """
    Test flattening, freezing, and thawing of models which contain parent (non-leaf) modules with parameters
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -208,6 +208,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 
 ### Fixed
+- Fixed `BaseFinetuning` callback to properly handle parent modules w/ parameters ([#7931](https://github.com/PyTorchLightning/pytorch-lightning/pull/7931))
```
Comment (carmocca) [inline CHANGELOG.md:211]:
```suggestion


- Fixed `BaseFinetuning` callback to properly handle parent modules w/ parameters ([#7931](https://github.com/PyTorchLightning/pytorch-lightning/pull/7931))

```

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -211,12 +211,25 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+
+- Fixed `apply_to_collection` works on Custom Collections now ([#7851](https://github.com/PyTorchLightning/pytorch-lightning/pull/7851))
+
+
 - Fixed ambiguous warning when both overfit and train dataloader shuffling are enabled ([#7685](https://github.com/PyTorchLightning/pytorch-lightning/pull/7685))
 
 
+- Fixed dataloaders are not reset when tuning the model ([#7566](https://github.com/PyTorchLightning/pytorch-lightning/pull/7566))
+
+
```
Comment (carmocca) [inline CHANGELOG.md:232]: Bad merge from somebody

```suggestion
- Fixed ambiguous warning when both overfit and train dataloader shuffling are enabled ([#7685](https://github.com/PyTorchLightning/pytorch-lightning/pull/7685))


```

**Review by tchaton — APPROVED** [review]
Good catch ! Thanks for using the BaseFinetuning Callback and catching this bug.

**Review by justusschock — APPROVED** [review]
